drivers/meter/ade7758/ade7758.c

```diff
@@ -0,0 +1,958 @@
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice,
+*    this list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* 3. Neither the name of Analog Devices, Inc. nor the names of its
+*    contributors may be used to endorse or promote products derived from this
+*    software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES, INC. “AS IS” AND ANY EXPRESS OR
+* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+* EVENT SHALL ANALOG DEVICES, INC. BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+* OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+* LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+* EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+#include "ade7758.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include "no_os_util.h"
+#include "no_os_spi.h"
+#include "no_os_gpio.h"
+#include "no_os_irq.h"
+#include "no_os_delay.h"
+#include "no_os_units.h"
+#include "no_os_alloc.h"
+#include "no_os_crc16.h"
+#include "no_os_print_log.h"
+#include <stdlib.h>
+#include <errno.h>
+#include <math.h>
+
+/**
+ * @brief Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+
+int ade7758_init(struct ade7758_dev **device,
+		 struct ade7758_init_param init_param)
+{
+	int ret;
+	/* ADE7758 device structure */
+	struct ade7758_dev *dev;
+	/* value read from register */
+	uint32_t reg_val;
+	/* timeout value */
+	uint16_t timeout = 1000;
+
+	dev = (struct ade7758_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI Initialization */
+	ret = no_os_spi_init(&dev->spi_desc,
+			     init_param.spi_init);
+	if (ret)
+		goto error_spi;
+
+	/* interrupt */
+	dev->irq_ctrl = init_param.irq_ctrl;
+	/* software reset */
+	ret = ade7758_sw_reset(dev);
+	if (ret)
+		goto error_spi;
+	/* wait for reset ack */
+	do {
+		ret = ade7758_get_int_status(dev, ADE7758_RESET_MSK, &reg_val);
+		if (ret)
+			goto error_spi;
+		timeout--;
+	} while ((!reg_val) && (timeout > 0));
+	if (timeout <= 0) {
```

Comment:
is there a case in which timeout will be < 0? I think this code block can be optimized a bit:
```
/* wait for reset ack */
do {
	ret = ade7758_get_int_status(dev, ADE7758_RESET_MSK, &reg_val);
	if (ret)
		goto error_spi;
} while ((!reg_val) && (timeout--));
if (!timeout) {
	ret = -ETIMEDOUT;
	goto error_spi;
}
```